odohlistener.go

```diff
@@ -0,0 +1,250 @@
+	TLSConfig *tls.Config
+}
+
+var _ Listener = &ODoHListener{}
+
+// NewODoHListener returns an instance of an oblivious DNS-over-HTTPS listener.
+func NewODoHListener(id, addr string, opt ODoHListenerOptions, resolver Resolver) (*ODoHListener, error) {
+	keyPair, err := getKeyPair(opt)
+	if err != nil {
+		Log.Error("Failed to generate HPKE key pair", "error", err)
+		return nil, err
+	}
+
+	dohOpt := DoHListenerOptions{
+		TLSConfig: opt.TLSConfig,
+		isChild:   true,
+	}
+	dohListen, err := NewDoHListener(id, addr, dohOpt, resolver)
+	if err != nil {
+		Log.Error("Failed to spawn DoH listener", "error", err)
+		return nil, err
+	}
+
+	l := &ODoHListener{
+		id:          id,
+		addr:        addr,
+		r:           resolver,
+		opt:         opt,
+		proxyClient: &http.Client{},
+		odohKeyPair: keyPair,
+		doh:         dohListen,
+	}
+
+	switch opt.OdohMode {
+	case "proxy":
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+	case "target":
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	default:
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	}
+	return l, nil
+}
+
+func (s *ODoHListener) Start() error {
+	return s.doh.Start()
+}
+
+func (s *ODoHListener) Stop() error {
+	return s.doh.Stop()
+}
+
+func (s *ODoHListener) ODoHproxyHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "only POST allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	contentType := r.Header.Get("Content-Type")
+	host := r.URL.Query().Get("targethost")
+	if host == "" {
+		http.Error(w, "no targethost specified", http.StatusBadRequest)
+		return
+	}
+	path := r.URL.Query().Get("targetpath")
+	if path == "" {
+		http.Error(w, "no targetpath specified", http.StatusBadRequest)
+		return
+	}
+
+	defer r.Body.Close()
+	b, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	Log.Debug("forwarding query to ODoH target",
+		slog.String("client", r.RemoteAddr),
+		slog.String("target", host),
+	)
+	response, err := forwardProxyRequest(s.proxyClient, host, path, b, contentType)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	if response.StatusCode != 200 {
+		http.Error(w, http.StatusText(response.StatusCode), response.StatusCode)
+		return
+	}
+
+	defer response.Body.Close()
+	rb, err := io.ReadAll(response.Body)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", contentType)
+	w.Write(rb)
+}
+
+func forwardProxyRequest(client *http.Client, targethost string, targetPath string, body []byte, contentType string) (*http.Response, error) {
+	targetURL := "https://" + targethost + targetPath
```

Comment:
This does look risky, and if got flagged by automatic code-scanning below. I don't remember how abuse is prevented in the RFC (will have to read it again). Maybe there's supposed to be an allowlist of target hosts? At least we should do more that trust all values and just concatenating into a URL. Perhaps:
- Validate the host is a valid hostname
- hostname, path length checks (limits)
- check against an allowlist?
- check the path to see if it's valid
- Build the URL with an appropriate library (https://pkg.go.dev/net/url) rather than string
operations
- ...